src/api.js

```diff
@@ -373,16 +382,42 @@ var remoteAuthentication = function(options, entry) {
 
   return function(req, res, next) {
     return authenticate(req).then(function(result) {
+      // Validate request hash if one is provided
+      if (typeof(result.hash) === 'string' && result.scheme === 'hawk') {
+        var hash = hawk.crypto.calculatePayloadHash(
+          new Buffer(req.text, 'utf-8'),
+          'sha256',
+          req.headers['content-type']
+        );
+        if (!cryptiles.fixedTimeComparison(result.hash, hash)) {
+          // create a fake auth-failed result with the failed hash
+          result = {
+            status: 'auth-failed',
+            message:
+              "Invalid payload hash: {{hash}}\n" +
+              "Computed payload hash: {{computedHash}}\n" +
+              "This happens when your request carries a signed hash of the " +
+              "payload and the hash doesn't match the hash we've computed " +
+              "on the server-side.",
+            computedHash: hash,
+          };
+        }
+      }
+
       /** Create method that returns list of scopes the caller has */
       req.scopes = function() {
-        // Check that we satisfy [], this ensures that payload hash is checked
-        // first... Just in case...
-        if(req.satisfies([[]], true)) {
-          return Promise.resolve(result.scopes || []);
+        if (result.status !== 'auth-success') {
+          return Promise.resolve([]);
         }
-        return Promise.resolve([]);
+        return Promise.resolve(result.scopes || []);
       };
 
+      if (result.status === 'auth-success') {
+        req.clientId = result.clientId || '(unknown)';
+      } else {
+        req.clientId = '(' + result.status + ')';
+      }
```

Comment:
Can we make this a function that returns a promise...
I did this with `scopes()` specifically, so that we could, in some future, make `authenticate` something we do when it's requested... Rather than doing it up-front for all API calls.